<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Data Science Workshop</title>
  <meta name="description" content="Lessons for the BSPR 2018 Data Science Workshop">
  <meta name="generator" content="bookdown 0.7 and GitBook 2.6.7">

  <meta property="og:title" content="Data Science Workshop" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Lessons for the BSPR 2018 Data Science Workshop" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Data Science Workshop" />
  
  <meta name="twitter:description" content="Lessons for the BSPR 2018 Data Science Workshop" />
  

<meta name="author" content="Alistair Bailey">



  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="tidyverse.html">
<link rel="next" href="fold-change-and-t-test.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Overview</a><ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#requirements"><i class="fa fa-check"></i>Requirements</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="intro.html"><a href="intro.html#what-are-r-and-rstudio"><i class="fa fa-check"></i><b>1.1</b> What are R and RStudio?</a><ul>
<li class="chapter" data-level="1.1.1" data-path="intro.html"><a href="intro.html#environments"><i class="fa fa-check"></i><b>1.1.1</b> Environments</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="intro.html"><a href="intro.html#why-learn-r-or-any-language"><i class="fa fa-check"></i><b>1.2</b> Why learn R, or any language ?</a></li>
<li class="chapter" data-level="1.3" data-path="intro.html"><a href="intro.html#finding-your-way-around-rstudio"><i class="fa fa-check"></i><b>1.3</b> Finding your way around RStudio</a><ul>
<li class="chapter" data-level="1.3.1" data-path="intro.html"><a href="intro.html#what-is-real"><i class="fa fa-check"></i><b>1.3.1</b> What is real?</a></li>
</ul></li>
<li class="chapter" data-level="1.4" data-path="intro.html"><a href="intro.html#where-am-i"><i class="fa fa-check"></i><b>1.4</b> Where am I?</a></li>
<li class="chapter" data-level="1.5" data-path="intro.html"><a href="intro.html#r-projects"><i class="fa fa-check"></i><b>1.5</b> R projects</a></li>
<li class="chapter" data-level="1.6" data-path="intro.html"><a href="intro.html#names"><i class="fa fa-check"></i><b>1.6</b> Naming things</a></li>
<li class="chapter" data-level="1.7" data-path="intro.html"><a href="intro.html#seeking-help"><i class="fa fa-check"></i><b>1.7</b> Seeking help</a><ul>
<li class="chapter" data-level="1.7.1" data-path="intro.html"><a href="intro.html#asking-for-help"><i class="fa fa-check"></i><b>1.7.1</b> Asking for help</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="tidyverse.html"><a href="tidyverse.html"><i class="fa fa-check"></i><b>2</b> Getting started in R and the tidyverse</a><ul>
<li class="chapter" data-level="2.1" data-path="tidyverse.html"><a href="tidyverse.html#tidy-data-and-the-tidyverse"><i class="fa fa-check"></i><b>2.1</b> Tidy data and the tidyverse</a></li>
<li class="chapter" data-level="2.2" data-path="tidyverse.html"><a href="tidyverse.html#data-visualisation"><i class="fa fa-check"></i><b>2.2</b> Data visualisation</a></li>
<li class="chapter" data-level="2.3" data-path="tidyverse.html"><a href="tidyverse.html#workflow-basics"><i class="fa fa-check"></i><b>2.3</b> Workflow basics</a><ul>
<li class="chapter" data-level="2.3.1" data-path="tidyverse.html"><a href="tidyverse.html#assigning-objects"><i class="fa fa-check"></i><b>2.3.1</b> Assigning objects</a></li>
<li class="chapter" data-level="2.3.2" data-path="tidyverse.html"><a href="tidyverse.html#calling-functions"><i class="fa fa-check"></i><b>2.3.2</b> Calling functions</a></li>
<li class="chapter" data-level="2.3.3" data-path="tidyverse.html"><a href="tidyverse.html#atomic-vectors"><i class="fa fa-check"></i><b>2.3.3</b> Atomic vectors</a></li>
<li class="chapter" data-level="2.3.4" data-path="tidyverse.html"><a href="tidyverse.html#attributes"><i class="fa fa-check"></i><b>2.3.4</b> Attributes</a></li>
<li class="chapter" data-level="2.3.5" data-path="tidyverse.html"><a href="tidyverse.html#factors"><i class="fa fa-check"></i><b>2.3.5</b> Factors</a></li>
<li class="chapter" data-level="2.3.6" data-path="tidyverse.html"><a href="tidyverse.html#lists"><i class="fa fa-check"></i><b>2.3.6</b> Lists</a></li>
<li class="chapter" data-level="2.3.7" data-path="tidyverse.html"><a href="tidyverse.html#matrices-and-arrays"><i class="fa fa-check"></i><b>2.3.7</b> Matrices and arrays</a></li>
<li class="chapter" data-level="2.3.8" data-path="tidyverse.html"><a href="tidyverse.html#data-frames"><i class="fa fa-check"></i><b>2.3.8</b> Data frames</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="tidyverse.html"><a href="tidyverse.html#learning-more-r"><i class="fa fa-check"></i><b>2.4</b> Learning more R</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="transform.html"><a href="transform.html"><i class="fa fa-check"></i><b>3</b> Importing and transforming data</a><ul>
<li class="chapter" data-level="3.1" data-path="transform.html"><a href="transform.html#some-definitions"><i class="fa fa-check"></i><b>3.1</b> Some definitions</a></li>
<li class="chapter" data-level="3.2" data-path="transform.html"><a href="transform.html#using-scripts"><i class="fa fa-check"></i><b>3.2</b> Using scripts</a></li>
<li class="chapter" data-level="3.3" data-path="transform.html"><a href="transform.html#running-code"><i class="fa fa-check"></i><b>3.3</b> Running code</a></li>
<li class="chapter" data-level="3.4" data-path="transform.html"><a href="transform.html#creating-a-r-script"><i class="fa fa-check"></i><b>3.4</b> Creating a R script</a></li>
<li class="chapter" data-level="3.5" data-path="transform.html"><a href="transform.html#setting-up-our-environment"><i class="fa fa-check"></i><b>3.5</b> Setting up our environment</a><ul>
<li class="chapter" data-level="3.5.1" data-path="transform.html"><a href="transform.html#bioconductor"><i class="fa fa-check"></i><b>3.5.1</b> Bioconductor</a></li>
</ul></li>
<li class="chapter" data-level="3.6" data-path="transform.html"><a href="transform.html#importing-data"><i class="fa fa-check"></i><b>3.6</b> Importing data</a></li>
<li class="chapter" data-level="3.7" data-path="transform.html"><a href="transform.html#calculating-fold-change-and-enrichment"><i class="fa fa-check"></i><b>3.7</b> Calculating fold change and enrichment</a></li>
<li class="chapter" data-level="3.8" data-path="transform.html"><a href="transform.html#fold-change-and-log-fold-change"><i class="fa fa-check"></i><b>3.8</b> Fold change and log-fold change</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="fold-change-and-t-test.html"><a href="fold-change-and-t-test.html"><i class="fa fa-check"></i><b>4</b> Fold change and t-test</a><ul>
<li class="chapter" data-level="4.1" data-path="fold-change-and-t-test.html"><a href="fold-change-and-t-test.html#heatmap-transformation"><i class="fa fa-check"></i><b>4.1</b> Heatmap transformation</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="visualising-proteomics-data.html"><a href="visualising-proteomics-data.html"><i class="fa fa-check"></i><b>5</b> Visualising proteomics data</a><ul>
<li class="chapter" data-level="5.1" data-path="visualising-proteomics-data.html"><a href="visualising-proteomics-data.html#creating-a-volcano-plot"><i class="fa fa-check"></i><b>5.1</b> Creating a volcano plot</a></li>
<li class="chapter" data-level="5.2" data-path="visualising-proteomics-data.html"><a href="visualising-proteomics-data.html#creating-a-heatmap"><i class="fa fa-check"></i><b>5.2</b> Creating a heatmap</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="going-further.html"><a href="going-further.html"><i class="fa fa-check"></i><b>6</b> Going further</a><ul>
<li class="chapter" data-level="6.1" data-path="going-further.html"><a href="going-further.html#getting-help-and-joining-the-r-community"><i class="fa fa-check"></i><b>6.1</b> Getting help and joining the R community</a></li>
<li class="chapter" data-level="6.2" data-path="going-further.html"><a href="going-further.html#communication-creating-reports-presentations-and-websites"><i class="fa fa-check"></i><b>6.2</b> Communication: creating reports, presentations and websites</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Data Science Workshop</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="transform" class="section level1">
<h1><span class="header-section-number">Chapter 3</span> Importing and transforming data</h1>
<div id="some-definitions" class="section level2">
<h2><span class="header-section-number">3.1</span> Some definitions</h2>
<ul>
<li>Importing means getting data into our R environment by turning into into a R object that we can then manipulate. The raw data file remains unchanged.</li>
<li>Inspecting means looking at the dataset to understand what it contains.</li>
<li>Tidying refers to getting data into a consistent format that makes it easy to use in later steps.</li>
</ul>
<p>A note here is that we are focusing on rectangular data, the sort that comes in rows and columns such as in a spreadsheet. Lots of our data types exist, such as images, which are beyond the scope of this lesson, but can also be handled by R.</p>
</div>
<div id="using-scripts" class="section level2">
<h2><span class="header-section-number">3.2</span> Using scripts</h2>
<p>Using the console is useful, but as we build up a workflow, that is to say, wring code to:</p>
<ul>
<li>load packages</li>
<li>load data</li>
<li>explore the data</li>
<li>and output some results</li>
</ul>
<p>Then it’s much more useful to contain this in a script: a document of our code.</p>
<p>Why? When we write and save our code in scripts, we can re-use it, share it or edit it. But <strong>most importantly a script is a record</strong>.</p>
<p>Cmd/Ctrl + Shift + N will open a new script file up and you should see something like Figure <a href="transform.html#fig:script-pane">3.1</a> with the script editor pane open:</p>

<div class="figure" style="text-align: center"><span id="fig:script-pane"></span>
<img src="img/rstudio_screenshot_four_panes.PNG" alt="Rstudio with the script editor pane open." width="80%" />
<p class="caption">
Figure 3.1: Rstudio with the script editor pane open.
</p>
</div>
</div>
<div id="running-code" class="section level2">
<h2><span class="header-section-number">3.3</span> Running code</h2>
<p>We can run a highlighted portion of code in your script if you click the Run button at the top of the scripts pane as shown in Figure <a href="transform.html#fig:run-script">3.2</a>.</p>
<p>(ref:run-script)</p>
<div class="figure" style="text-align: center"><span id="fig:run-script"></span>
<img src="img/run_script.PNG" alt="(ref:run-script)" width="80%" />
<p class="caption">
Figure 3.2: (ref:run-script)
</p>
</div>
<p>You can run the entire script by clicking the Source button.</p>
<p>Or we can run chunks of code if we split our script into sections, see below.</p>
</div>
<div id="creating-a-r-script" class="section level2">
<h2><span class="header-section-number">3.4</span> Creating a R script</h2>
<p>We first need to create a script that will form the basis of our analysis.</p>
<p>Go to the file menu and select New Files &gt; R script. This should open the script editor pane.</p>
<p>Now let’s save the script, by going to File &gt; Save and we should find ourselves prompted to save the script in our Project Directory.</p>
<p>Following the advice about <a href="intro.html#names">naming things</a> we can create a new R script called <code>01-bspr-workshop-july-2018</code>.</p>
<p>This name is machine readable (no spaces or special characters), human readable, and works well with default ordering by beginning with <code>01</code>.</p>
</div>
<div id="setting-up-our-environment" class="section level2">
<h2><span class="header-section-number">3.5</span> Setting up our environment</h2>
<p>At the head of our script it’s common to put a title, the name of the author and the date, and any other useful information. This is created as comments using the <code>#</code> at the start of each line.</p>
<p>It’s then usual to follow this by code to load the packages we need into our our R environment using the <code>library()</code> function and providing the name of the package we wish to load. Packages are collections of R functions.</p>
<p>Often we break the code up into regions by adding dashes (or equals symbols) to the comment line. This enables us to run chunks of the script separately from running the whole script when using our code.</p>
<p>Here is a typical head for a script:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># My workshop script</span>
<span class="co"># 7th July 2018</span>
<span class="co"># Alistair Bailey</span>

<span class="co"># Load packages ----------------------------------------------------------------</span>
<span class="kw">library</span>(tidyverse)
<span class="kw">library</span>(janitor)
<span class="kw">library</span>(gplots)
<span class="kw">library</span>(pheatmap)</code></pre></div>
<div id="bioconductor" class="section level3">
<h3><span class="header-section-number">3.5.1</span> Bioconductor</h3>
<p>As an aside there are many proteomics specific R packages, these are generally found through <a href="https://www.bioconductor.org/">Bioconductor</a> which is a project that was initiated in 2001 to create tools for the analysis of high-throughput genomic data, but also includes other ’omics data tools <span class="citation">(R. C. Gentleman et al. <a href="#ref-gentleman2004">2004</a>,<span class="citation">Huber et al. (<a href="#ref-huber2015">2015</a>)</span>)</span>.</p>
<p>Exploring Bioconductor is beyond our scope here, but well worth exploring for manipulation and analysis of raw data formats such as mzxml files.</p>
</div>
</div>
<div id="importing-data" class="section level2">
<h2><span class="header-section-number">3.6</span> Importing data</h2>
<p>Assuming our data is in a flat format, we can import it into our environment using the tidyverse <code>readr</code> package. A flat format is a file that contains only text such as <code>csv</code> or tab separated data without meta-data (extra information), such as the colour of cells in an excel file.</p>
<p>If we our data was an excel file, we can use the tidyverse <code>readxl</code> package to import the data.</p>
<p>For the purposes of this workshop we have a <code>csv</code> (comma separated variable) file.</p>
<p>The example data can be downloaded from <a href="link">here</a> as saved to our project directory. Check the <code>Files</code> pane to see it’s there.</p>
<p>We then import data and assign it to an object we’ll call <code>data</code> like so:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Import example data ----------------------------------------------------------</span>

<span class="co"># Import the example data with read_csv from the readr package</span>
<span class="co"># Tell read_csv that missing values are the string NaN</span>
<span class="co"># Clean up the variable names with clean_names() from the janitor package</span>
dat &lt;-<span class="st"> </span>readr<span class="op">::</span><span class="kw">read_csv</span>(<span class="st">&quot;070718-proteomics-example-data.csv&quot;</span>,
                <span class="dt">na =</span> <span class="st">&quot;NaN&quot;</span>) <span class="op">%&gt;%</span>
<span class="st">                </span>janitor<span class="op">::</span><span class="kw">clean_names</span>()</code></pre></div>
</div>
<div id="calculating-fold-change-and-enrichment" class="section level2">
<h2><span class="header-section-number">3.7</span> Calculating fold change and enrichment</h2>
</div>
<div id="fold-change-and-log-fold-change" class="section level2">
<h2><span class="header-section-number">3.8</span> Fold change and log-fold change</h2>
<p>Fold changes are ratios, the ratio of say protein expression before and after treatment, where a value larger than 1 for a protein implies that protein expression was greater after the treatment.</p>
<p>In life sciences, fold change is often reported as log-fold change. Why is that? There are at least two reasons which can be shown by plotting.</p>
<p>One is that ratios are not symmetrical around 1, so it’s difficult to observe both changes in the forwards and backwards direcion i.e. proteins where expression went up and proteins where expression went down due to treatment. When we transform ratios on a log scale, the scale becomes symmetric around 0 and thus we can now observe the distribution of ratios in terms of positive, negative or no change.</p>

<div class="figure"><span id="fig:fold-change-1"></span>
<img src="bspr-workshop-2018_files/figure-html/fold-change-1-1.png" alt="Ratios are not symmetric around one, logratios are symmetric around zero." width="672" />
<p class="caption">
Figure 3.3: Ratios are not symmetric around one, logratios are symmetric around zero.
</p>
</div>
<p>A second reason is that transforming values onto a log scale changes where the numbers actually occur when plotted on that scale. If we consider the log scale to represent magnitudes, then we can more easily see changes of small and large magnitudes when we plot the data.</p>
<p>For example, a fold change of 32 times can be either a ratio 1/32 or 32/1.</p>
<p>As shown in Figure <a href="transform.html#fig:fold-change-2">3.4</a>, 1/32 is much closer to 1 than 32/1, but transformed to a log scale we see that in terms of magnitude of difference it is the same as 32/1.</p>

<div class="figure"><span id="fig:fold-change-2"></span>
<img src="bspr-workshop-2018_files/figure-html/fold-change-2-1.png" alt="Transformation of scales using log transformation." width="672" />
<p class="caption">
Figure 3.4: Transformation of scales using log transformation.
</p>
</div>
</div>
</div>
<h3>References</h3>
<div id="refs" class="references">
<div id="ref-gentleman2004">
<p>Gentleman, Robert C, Vincent J Carey, Douglas M Bates, Ben Bolstad, Marcel Dettling, Sandrine Dudoit, Byron Ellis, et al. 2004. “Bioconductor: Open Software Development for Computational Biology and Bioinformatics.” <em>Genome Biology</em> 5 (10): R80. doi:<a href="https://doi.org/10.1186/gb-2004-5-10-r80">10.1186/gb-2004-5-10-r80</a>.</p>
</div>
<div id="ref-huber2015">
<p>Huber, Wolfgang, Vincent J Carey, Robert Gentleman, Simon Anders, Marc Carlson, Benilton S Carvalho, Hector Corrada Bravo, et al. 2015. “Orchestrating High-Throughput Genomic Analysis with Bioconductor.” <em>Nature Methods</em> 12 (2): 115–21. doi:<a href="https://doi.org/10.1038/nmeth.3252">10.1038/nmeth.3252</a>.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="tidyverse.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="fold-change-and-t-test.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": ["bspr-workshop-2018.pdf", "bspr-workshop-2018.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
